File: 0228-summary-ranges/solution.go
func summaryRanges(nums []int) []string {
    if len(nums) == 0 {
        return []string{}
    }

    ranges := make([]string, 0)

    left := 0
    for right := left + 1; right < len(nums); right++ {
        if nums[right] - nums[right - 1] > 1 {
            // build range
            if left != right - 1 {
                ranges = append(ranges, fmt.Sprintf("%d->%d", nums[left], nums[right - 1]))
            } else {
                ranges = append(ranges, fmt.Sprintf("%d", nums[left]))
            }
            left = right
            continue
        }
    }

    if left < len(nums) - 1 {
        ranges = append(ranges, fmt.Sprintf("%d->%d", nums[left], nums[len(nums) - 1]))
    } else {
        ranges = append(ranges, fmt.Sprintf("%d", nums[left]))
    }

    return ranges
}

File: 0228-summary-ranges/solution.js
/**
 * @param {number[]} nums
 * @return {string[]}
 */
var summaryRanges = function(nums) {
    if (nums.length === 0) return [];

    let result = [];
    let leftPtr = 0;
    for (let rightPtr = leftPtr + 1; rightPtr < nums.length; rightPtr++) {
        if (nums[rightPtr] - nums[rightPtr - 1] > 1) { // stop the range
            if (leftPtr !== rightPtr - 1) {
                result.push(`${nums[leftPtr]}->${nums[rightPtr - 1]}`);
            } else {
                result.push(`${nums[leftPtr]}`);
            }
            leftPtr = rightPtr;
            continue;
        }
    }

    if (leftPtr < nums.length - 1) { // left is not the latest element
        result.push(`${nums[leftPtr]}->${nums[nums.length - 1]}`);
    } else {
        result.push(`${nums[leftPtr]}`);
    }

    return result;
};
